### PR TITLE
 Refactor egress IP assignment so that all operations are done as a single libovsdb transaction

### PR DESCRIPTION
**Pull request by kyrtapz**
Refactor egress IP assignment so that all operations are done as a single libovsdb transaction.

Commit egress IP pod assignment as one transaction with the following ops:
 1. Add SNAT towards egress IP on egress gw router
 2. Add a router policy to forward pod traffic to the egress node
 3. Delete SNAT towards nodeIP(only if the egress node is the same as the pod host)

Commit egress IP pod assignment removal as one transaction with the following ops:
 1. Add SNAT towards nodeIP(only if the egress node is the same as the pod host)
 2. Remove router policy that forwards pod traffic to the egress node
 3. Remove SNAT towards egress IP from the egress node

NOTE: Unfortunately there is still a chance that the flow won't be configured when the packet arrives at the egress node but it is less likely to happen with the proposed change.

/cc @dceara 

Signed-off-by: Patryk Diak <pdiak@redhat.com>

### CODE REVIEW

File: go-controller/pkg/ovn/egressip.go
Comment (trozet) [inline go-controller/pkg/ovn/egressip.go:2193]: why dont we just make this whole thing more efficient and do it as a single transaction? That way either the SNAT and the reroute policy get configured at the same time, or if an error happens neither of them are configured?

**Review by trozet — CHANGES_REQUESTED** [review]

File: go-controller/pkg/ovn/egressip.go
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go:2193]: Agreed, good idea.
I will work on refactoring the addition/deletion so they are always done as a single transaction.

Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go]: @trozet @tssurya done

File: go-controller/pkg/ovn/egressip.go
Comment (tssurya) [inline go-controller/pkg/ovn/egressip.go:2194]: self-note: 1) we are deleting SNAT towards nodeIP for the pod 2) we are adding SNAT towards egressIP of the pod 3) we are adding rereoute policy for the pod

nit: would be nice to outline in the commit message (and update PR description so that anyone who doesn't have context can understand what ops you are talking about) the fact that these are the 3 ops we are combining together and that adding SNAT will be done before adding reroute to avoid that small air gap. - this will set stage for your other PR as well

**Review by tssurya — COMMENTED** [review]
overall LGTM, liked how you split up your commits, they made my review life easy, mostly some nits, I'll leave it to your judgement if you want to fox them.

Mainly question around should we combine `addExternalGWPodSNAT` or not? see comments inline.

File: go-controller/pkg/ovn/egressip.go
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go:2295]: I saw that method but it does return pretty values used in logging, additionally ovn expects `ip4/ip6` instead of `ipv6/ipv4`?

File: go-controller/pkg/ovn/egressip.go
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go]: changed it to:

> it means the pod host is different from the egressNode that is managing the pod

File: go-controller/pkg/libovsdbops/router.go
Comment (kyrtapz) [reply go-controller/pkg/libovsdbops/router.go:363]: I think it would be fairly tricky to properly to clearly explain it...
I followed the same approach used in other `...Ops(...` methods

File: go-controller/pkg/ovn/egressip.go
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go:2224]: I added that to `deletePodEgressIPAssignment` so it is consistent with `addPodEgressIPAssignment`.
I don't think we should do the `deleteEgressIPAssignments` in this PR as it might results in hundreds of ops per txn, this is something we could explore separately.

Comment (tssurya) [reply go-controller/pkg/ovn/egressip.go:2224]: sounds good yes!

File: go-controller/pkg/ovn/egressgw.go
Comment (flavio-fernandes) [inline go-controller/pkg/ovn/egressgw.go:579]: Use %w instead of %v

**Review by flavio-fernandes — COMMENTED** [review]
short of a little nit...
/lgtm

File: go-controller/pkg/ovn/egressip.go
Comment (kyrtapz) [reply go-controller/pkg/ovn/egressip.go:2221]: If we want to refactor the old code so we always return errors wrapped with `%w` I think we could do a repo wide refactor in a separate pr.

**Review by tssurya — APPROVED** [review]
/lgtm
on the new changes since last review, 
@kyrtapz : seems like you need to rebase?

**Comment by trozet** [comment]
re-running failed jobs before merge

**Pull request merged**
